RenderStateNotation/src/RenderStateNotationParserImpl.cpp

```diff
@@ -335,8 +335,50 @@ Bool RenderStateNotationParserImpl::ParseString(const Char* StrData, Uint32 Leng
                     if (ShaderType != SHADER_TYPE_UNKNOWN)
                         ResourceDesc.Desc.ShaderType = ShaderType;
 
-                    m_ShaderNames.emplace(HashMapStringKey{ResourceDesc.Desc.Name, false}, StaticCast<Uint32>(m_Shaders.size()));
-                    m_Shaders.push_back(ResourceDesc);
+                    auto const Iter = m_ShaderNames.find(ResourceDesc.Desc.Name);
+                    if (Iter != m_ShaderNames.end())
+                    {
+                        auto CompareMacros = [](const ShaderMacro* pLHS, const ShaderMacro* pRHS) //
+                        {
+                            if ((pLHS == nullptr) != (pRHS == nullptr))
+                                return false;
+                            if (pLHS == pRHS)
+                                return true;
+
+                            VERIFY_EXPR(pLHS != nullptr && pRHS != nullptr);
+                            while (!(*pLHS == ShaderMacro{} || *pRHS == ShaderMacro{}) && *pLHS == *pRHS)
+                            {
+                                ++pLHS;
+                                ++pRHS;
+                            }
+                            return *pLHS == ShaderMacro{} && *pRHS == ShaderMacro{};
+                        };
+
+                        auto CompareShaderCI = [&CompareMacros](const ShaderCreateInfo& LHS, const ShaderCreateInfo& RHS) //
+                        {
+                            return LHS.Desc == RHS.Desc &&
+                                LHS.SourceLanguage == RHS.SourceLanguage &&
+                                LHS.HLSLVersion == RHS.HLSLVersion &&
+                                LHS.GLSLVersion == RHS.GLSLVersion &&
+                                LHS.GLESSLVersion == RHS.GLESSLVersion &&
+                                LHS.UseCombinedTextureSamplers == RHS.UseCombinedTextureSamplers &&
+                                LHS.CompileFlags == RHS.CompileFlags &&
+                                LHS.ShaderCompiler == RHS.ShaderCompiler &&
+                                SafeStrEqual(LHS.EntryPoint, RHS.EntryPoint) &&
+                                SafeStrEqual(LHS.FilePath, RHS.FilePath) &&
+                                SafeStrEqual(LHS.CombinedSamplerSuffix, RHS.CombinedSamplerSuffix) &&
+                                CompareMacros(LHS.Macros, RHS.Macros);
+                        };
```

Comment:
It is better to move this to core to `ShaderCreateInfo::operator==`

Reply:
We can't use a generic comparator because it will only work in this case